File: domain/src/main/java/org/openmedstack/eventstore/DelegateSagaRepository.java
package org.openmedstack.eventstore;

import openmedstack.IProvideTenant;
import org.openmedstack.domain.HandlerForDomainEventNotFoundException;
import org.openmedstack.domain.Saga;

import java.util.HashMap;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

public class DelegateSagaRepository implements SagaRepository {
    private final IProvideTenant _tenantId;
    private final IStoreEvents _eventStore;
    private final IConstructSagas _factory;

    public DelegateSagaRepository(IProvideTenant tenantId, IStoreEvents eventStore, IConstructSagas factory)
    {
        _tenantId = tenantId;
        _eventStore = eventStore;
        _factory = factory;
    }

    @Override
    public <TSaga extends Saga> CompletableFuture<TSaga> getById(Class<TSaga> type,String sagaId) {
        return openStream(_tenantId.getTenantName(), sagaId, Integer.MAX_VALUE).thenApplyAsync(stream -> {
            try {
                return buildSaga(type, sagaId, stream);
            } catch (HandlerForDomainEventNotFoundException e) {
                throw new RuntimeException(e);
            }
        });
    }

    @Override
    public CompletableFuture<Boolean> save(Saga saga, Consumer<HashMap<String, Object>> updateHeaders) {
        if (saga == null) {
            throw new IllegalArgumentException("Saga cannot be null");
        }
        var commitId = UUID.randomUUID();
        var headers = PrepareHeaders(saga, updateHeaders);
        return prepareStream(_tenantId.getTenantName(), saga, headers).thenApplyAsync(eventStream -> {
            try {
                persist(eventStream, commitId);
                saga.clearUncommittedEvents();
                saga.clearUndispatchedMessages();
                return true;
            } catch (ConcurrencyException e) {
                return false;
            }
        });
    }

    private CompletableFuture<IEventStream> openStream(String bucketId, String sagaId, int maxVersion)
    {
        CompletableFuture<IEventStream> eventStream;
        try
        {
            eventStream = _eventStore.openStream(bucketId, sagaId, 0, maxVersion);
        }
        catch (StreamNotFoundException)
        {
            eventStream =  _eventStore.createStream(bucketId, sagaId);
        }
        return eventStream;
    }

    private <TSaga extends Saga> TSaga buildSaga(Class<TSaga> type, String sagaId, IEventStream stream) throws HandlerForDomainEventNotFoundException
    {
        var saga = (TSaga)_factory.build(type, sagaId);
        for (var x : stream.getCommittedEvents()){
            saga.transition(x.getBody());
    }

        saga.clearUncommittedEvents();
        saga.clearUndispatchedMessages();
        return saga;
    }

    private static HashMap<String, Object> PrepareHeaders(Saga saga, Consumer<HashMap<String, Object>> updateHeaders)
    {
        var dictionary = new HashMap<String, Object>();
        dictionary.put("SagaType", saga.getClass().getTypeName());
    if(updateHeaders!= null){
        updateHeaders.accept(dictionary);
    }
        int num = 0;
        for (var undispatchedMessage : saga.getUndispatchedMessages())
        {
            dictionary.put("UndispatchedMessage." + num++, undispatchedMessage);
        }
        return dictionary;
    }

    private CompletableFuture<IEventStream> prepareStream(String bucketId, Saga saga, HashMap<String, Object> headers) {
        return openStream(bucketId, saga.getId(), saga.getVersion())
                .thenApply(stream -> {

                    for (var header : headers.entrySet()) {
                        stream.getUncommittedHeaders().put(header.getKey(), header.getValue());
                    }

                    saga.getUncommittedEvents().forEach(o -> {
                        stream.add(new EventMessage(o, null));
                    });

                    return stream;
                });
    }

    private CompletableFuture persist(IEventStream stream, UUID commitId) throws ConcurrencyException
    {
        try
        {
            return stream.commitChanges(commitId);
        }
        catch (DuplicateCommitException ex)
        {
            stream.clearChanges();
        }
        return CompletableFuture.completedFuture(true);
    }
}

File: domain/src/main/kotlin/org/openmedstack/eventstore/DelegateSagaRepository.kt
package org.openmedstack.eventstore

import org.openmedstack.IProvideTenant
import org.openmedstack.domain.HandlerForDomainEventNotFoundException
import org.openmedstack.domain.Saga
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.function.Consumer
import kotlin.collections.HashMap

class DelegateSagaRepository(private val _tenantId: IProvideTenant, private val _eventStore: IStoreEvents, private val _factory: IConstructSagas) : SagaRepository {
    override fun <TSaga : Saga> getById(type: Class<TSaga>, sagaId: String): CompletableFuture<TSaga> {
        return openStream(_tenantId.tenantName, sagaId, Int.MAX_VALUE).thenApplyAsync { stream: IEventStream ->
            try {
                return@thenApplyAsync buildSaga(type, sagaId, stream)
            } catch (e: HandlerForDomainEventNotFoundException) {
                throw RuntimeException(e)
            }
        }
    }

    override fun save(saga: Saga, updateHeaders: Consumer<HashMap<String, Any>>): CompletableFuture<Boolean> {
        val commitId = UUID.randomUUID()
        val headers = prepareHeaders(saga, updateHeaders)
        return prepareStream(_tenantId.tenantName, saga, headers).thenApplyAsync { eventStream: IEventStream ->
            try {
                persist(eventStream, commitId)
                saga.clearUncommittedEvents()
                saga.clearUndispatchedMessages()
                return@thenApplyAsync true
            } catch (e: ConcurrencyException) {
                return@thenApplyAsync false
            }
        }
    }

    private fun openStream(bucketId: String?, sagaId: String, maxVersion: Int): CompletableFuture<IEventStream> {
        return _eventStore.openStream(bucketId, sagaId, 0, maxVersion)
    }

    @Throws(HandlerForDomainEventNotFoundException::class)
    private fun <TSaga : Saga> buildSaga(type: Class<TSaga>, sagaId: String, stream: IEventStream): TSaga {
        val saga: TSaga = _factory.build(type, sagaId)
        for (x in stream.committedEvents) {
            saga.transition(x.body)
        }
        saga.clearUncommittedEvents()
        saga.clearUndispatchedMessages()
        return saga
    }

    private fun prepareStream(bucketId: String?, saga: Saga, headers: HashMap<String, Any>): CompletableFuture<IEventStream> {
        return openStream(bucketId, saga.id, saga.version)
                .thenApplyAsync { stream: IEventStream ->
                    for ((key, value) in headers) {
                        stream.uncommittedHeaders[key] = value
                    }
                    saga.uncommittedEvents.forEach(Consumer { o: Any -> stream.add(EventMessage(o, HashMap())) })
                    stream
                }
    }

    @Throws(ConcurrencyException::class)
    private fun persist(stream: IEventStream, commitId: UUID): CompletableFuture<*>? {
        try {
            return stream.commitChanges(commitId)
        } catch (ex: DuplicateCommitException) {
            stream.clearChanges()
        }
        return CompletableFuture.completedFuture(true)
    }

    companion object {
        private fun prepareHeaders(saga: Saga, updateHeaders: Consumer<HashMap<String, Any>>): HashMap<String, Any> {
            val dictionary = HashMap<String, Any>()
            dictionary["SagaType"] = saga.javaClass.typeName
            updateHeaders.accept(dictionary)
            var num = 0
            for (undispatchedMessage in saga.undispatchedMessages) {
                dictionary["UndispatchedMessage." + num++] = undispatchedMessage
            }
            return dictionary
        }
    }
}